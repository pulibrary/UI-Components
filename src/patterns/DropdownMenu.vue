<template>
  <component :is="type" :class="['lux-dropdown-menu']">
    <slot></slot>
  </component>
</template>

<script>
/**
 * Dropdowns allows a user to select a value from a series of options.
 */
export default {
  name: "DropdownMenu",
  status: "prototype",
  release: "1.0.0",
  type: "Pattern",
  props: {
    /**
     * The html element name used for the container
     */
    type: {
      type: String,
      default: "div",
    },
    /**
     * Style variation to give additional meaning.
     * `default, disabled, active`
     */
    align: {
      type: String,
      default: "left",
      validator: value => {
        return value.match(/(left|right)/)
      },
    },
  },
}
</script>

<style lang="scss" scoped>
.lux-dropdown-menu {
  @include reset;
  @include stack-space($space-small);
  margin: 0.25rem;
  font-family: $font-family-text;
  font-size: $font-size-base;
  line-height: 1;
  color: set-text-color($color-rich-black, $color-white);
  position: relative;
  display: inline-block;
  top: 0;

  .dropdown {
    position: relative;
    vertical-align: middle;
    line-height: 1;
    margin: 0;

    &:hover,
    &:focus {
      background: $color-grayscale-lighter;
    }
  }

  .menu,
  .nav {
    opacity: 0;
    transition: all 0s ease-in-out 1s;
    visibility: hidden;
  }

  &:hover > .menu,
  &:focus > .menu,
  &:focus-within > .menu,
  &:hover > .nav,
  &:focus > .nav,
  &:focus-within > .nav,
  .menu:hover,
  .nav:hover {
    opacity: 1;
    z-index: 9999;
    transition-delay: 0s;
    visibility: visible;
  }

  /deep/ .nav {
    background: $color-white;
    box-shadow: $box-shadow-small;

    position: absolute;
    min-width: calc(100% - 1px);
    margin: 1px 0 0;
    z-index: 9999;

    ul {
      list-style-type: none;
      margin: 0;
      padding: 0;
    }

    li {
      display: block;
    }

    .nav-item {
      background: $color-white;
      border: 0;
      color: $color-rich-black;
      cursor: pointer;
      display: block;
      padding: 0.5rem 1rem;
      margin: 0;
      width: 100%;
      text-align: left;
      font-size: $font-size-small;

      &:hover,
      &:focus {
        color: $color-rich-black;
        background: $color-grayscale-lighter;
        transition: opacity 0.1s ease;
      }

      &:active {
        transform: scale(0.99);
      }
    }

    .disabled {
      color: $color-grayscale-dark;

      &:hover,
      &:focus {
        background: $color-white;
        cursor: not-allowed;
        outline: none;
      }
    }
  }
}
</style>

<docs>
  ```jsx
  <dropdown-menu>
    <input-button variation="dropdown">Selection Options</input-button>
    <menu-bar type="menu" active="All" :menu-items="[
      {name: 'All', component: 'All'},
      {name: 'None', component: 'None'},
      {name: 'Alternate', component: 'Alternate', disabled: true},
      {name: 'Inverse', component: 'Inverse'}
    ]"/>
  </dropdown-menu>
  ```
</docs>
