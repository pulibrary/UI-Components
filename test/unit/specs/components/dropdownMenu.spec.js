import { createLocalVue, mount } from "@vue/test-utils"
import DropdownMenu from "@/patterns/DropdownMenu.vue"

// create an extended `Vue` constructor
const localVue = createLocalVue()

let wrapper

describe("DropdownMenu.vue", () => {
  beforeEach(() => {
    wrapper = mount(DropdownMenu, {
      localVue,
    })
  })

  it("should be of the element type that is passed in as a prop", () => {
    expect(wrapper.find("div").exists()).toBe(true)
    expect(wrapper.find("form").exists()).toBe(false)
    // can't setProps here because the component must be regenerated to change its wrapper
    const wrapper2 = mount(DropdownMenu, {
      localVue,
      propsData: {
        type: "form",
      },
    })
    expect(wrapper2.find("div").exists()).toBe(false)
    expect(wrapper2.find("form").exists()).toBe(true)
  })

  it("has the expected html structure", () => {
    expect(wrapper.element).toMatchSnapshot()
  })
})
