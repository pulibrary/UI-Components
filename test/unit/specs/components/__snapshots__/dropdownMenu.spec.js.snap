// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DropdownMenu.vue has the expected html structure 1`] = `
<div
  class="lux-dropdown-menu medium"
>
  <input-button-stub
    class="lux-dropdown-button"
    size="medium"
    variation="dropdown"
  >
    
  </input-button-stub>
   
  <menu-bar-stub
    active="All"
    class=""
    type="buttons"
  />
</div>
`;
